[USMON-1189] Add Postgres telemetry configuration

amitslavin opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

* Introduce `max_postgres_telemetry_buffer` as a configuration for setting the base postgres telemetry buffer. This will provide us with a better understanding of what the optimal value for the kernel buffer size should be.
The default value of the `max_postgres_telemetry_buffer` is set to `ebpf.BufferSize` which determines the postgres max buffer size.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

We [recently added](https://github.com/DataDog/datadog-agent/pull/26988) telemetry to the USM Postgres support. After analyzing the results, we found that most of the query lengths are greater than our maximum bucket range. To address this issue and understand what the best value for the PostgreSQL buffer size is, we want to be able to control the range we test.

[Link](https://ddstaging.datadoghq.com/s/c894ecec5/4v6-58h-yxp) to the Postgres internals dashboard, which includes the telemetry metrics for the query length buckets.


<img width="1652" alt="Screenshot 2024-06-27 at 13 42 49" src="https://github.com/DataDog/datadog-agent/assets/108348428/31207df3-bd1a-41d1-89e1-82d3c82e9a7c">


<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

amitslavin committed:
changed postgres telemetry base buffer size

amitslavin committed:
fixed bucket-1 range documentation

guyarb commented on pkg/network/protocols/postgres/telemetry.go:
```diff
@@ -15,14 +15,15 @@ import (
 )
 
 const (
+	telemetryBaseBufferSize                 = 160
```
Why not using the same value from the kernel?
Having "duplication" can result in having a fork in the future (a.k.a different values)

amitslavin replied:
I am starting only with telemetry as the changes in the kernel value are affecting the verifier and are more challenging. I plan to understand better what the value should be, and only then work on the kernel change. I do agree with you that this change can be tested to understand if the fixed size is correct (will start testing on staging). Regarding making it configurable, I am not sure how useful it would be, as when we find the correct size, it should be constant.

amitslavin replied:
I have separated it intentionally to distinguish between changing the value and understanding what the most suitable value is. Once I determine the value we plan to use, I will ensure that the values for telemetry and the kernel are equal.

guyarb replied:
then, I don't understand the goal  of the PR
if the branch is used only for testing purposes, you don't need a PR

amitslavin replied:
The purpose is to create a method to better understand the most appropriate buffer size. I have created a separate PR with a configuration change to allow for faster and more accurate understanding.

I can create a new image with a different value each time and deploy it to understand the impact. However, I'm not sure if this is better than having one PR, which would allow me to update the configuration more easily.

guyarb replied:
but you can use this branch for the testing until you get the results of the "optimal" buffer size
I'm curious why we need to merge it to main

guyarb reviewed (commented):
I'm wondering if the PR is only for testing purposes

Anyway, since everything is usermode, you can do the extra mile, and make the original value (telemetry buffer size) configurable

amitslavin committed:
Empty-Commit

amitslavin committed:
Added configuration variable for the telemetry buffer size

amitslavin commented:
> I'm wondering if the PR is only for testing purposes
> 
> Anyway, since everything is usermode, you can do the extra mile, and make the original value (telemetry buffer size) configurable

I agree that a configuration value can help in better understanding what the best value to use is. I have added a configuration value for this in the PR as well.

guyarb commented on pkg/network/config/config_test.go:
```diff
@@ -1246,6 +1246,26 @@ service_monitoring_config:
 	})
 }
 
+func TestMaxPostgresTelemetryBuffered(t *testing.T) {
```
add a test for the default value

guyarb commented on pkg/network/protocols/postgres/protocol.go:
```diff
@@ -200,7 +200,11 @@ func (p *protocol) processPostgres(events []EbpfEvent) {
 		eventWrapper := NewEventWrapper(tx)
 		p.statskeeper.Process(eventWrapper)
-		p.telemetry.Count(tx, eventWrapper)
+		if p.cfg.MaxPostgresTelemetryBuffer > 0 {
```
can you elaborate on the condition here?

amitslavin replied:
Actually, this is a bug. I have changed it to check if the cfg value is greater than the default value, and if so, I am using the cfg value.

amitslavin committed:
fixed cr notes

gjulianm reviewed (approved):
LGTM for ebpf-platform files

amitslavin committed:
Merge remote-tracking branch 'origin/main' into amit.slavin/increse-telmetry-base-size

# Conflicts:
#	pkg/config/setup/system_probe.go
#	pkg/network/config/config.go
#	pkg/network/protocols/postgres/telemetry.go
#	pkg/network/protocols/postgres/telemetry_test.go
#	pkg/network/tracer/tracer_windows.go

amitslavin committed:
changed postgres types location to new package

fix imports

add types_linux.go in the new pkg

fix uts using postgres types

fix import

fix typo

amitslavin committed:
fix copyright date

amitslavin committed:
fix cr notes

amitslavin committed:
changed ebpfpostgres to subpackage

amitslavin committed:
Merge remote-tracking branch 'origin/main' into amit.slavin/increse-telmetry-base-size

amitslavin committed:
Merge branch 'amit.slavin/relocate-postgres-types' into amit.slavin/increse-telmetry-base-size

amitslavin committed:
include latest changes

amitslavin committed:
Merge remote-tracking branch 'origin/main' into amit.slavin/increse-telmetry-base-size

# Conflicts:
#	pkg/network/protocols/postgres/telemetry.go
#	pkg/network/protocols/postgres/telemetry_test.go

amitslavin committed:
Merge remote-tracking branch 'origin/main' into amit.slavin/increse-telmetry-base-size

amitslavin committed:
fix ci

amitslavin committed:
validating the config from subpackage

amitslavin committed:
fix licence and go format

amitslavin committed:
Merge remote-tracking branch 'origin/main' into amit.slavin/increse-telmetry-base-size

amitslavin committed:
use config test to validate the default

amitslavin committed:
changed test function name

amitslavin committed:
added doc

amitslavin committed:
Merge remote-tracking branch 'origin/main' into amit.slavin/increse-telmetry-base-size

amitslavin committed:
removed MaxPostgresTelemetryBuffer from windows

amitslavin commented:
/merge
